src/pipe.rs

```diff
@@ -28,7 +29,7 @@ impl Drop for Pipe {
     fn drop(&mut self) {
         // This is the correct thing to do, but is not super important since the `Sink`
         // implementation of `Sender` just returns `Ok` without doing anything else.
-        let _ = self.dest.close();
+        let _ = block_on(self.dest.close());
```

Comment:
How about instead of the `block_on` here to make this compatible with this sync trait, you try to use [AsyncWrite](https://docs.rs/tokio/0.2.21/tokio/io/trait.AsyncWrite.html)? It's also re-exported as `actix_codec::AsyncWrite`. It might make this code more nicely integrated if all of this were async and then we wouldn't need the `block_on`. :)

Reply:
Same applies to the other `block_on`s down below.

Reply:
`Pipe` exists because `create_archive` only accepts `Write`ables.
If it accepted `AsyncWrite`, `Pipe` was not required because `Sender` itself implements `AsyncWrite` (via `Sink` trait).

I'll investigate if I can rewrite `create_archive` and its family so that they can accept `AsyncWrite`.

Reply:
That'd be great! I think it would result in less code and hopefully simpler code, too, given that there'd be no conversion between sync/async code.

Reply:
I found an async implementation of gzip: [async-compression](https://github.com/Nemo157/async-compression).
The bad news is that I couldn't find the one for tar and zip.
(There is [async-tar](https://github.com/dignifiedquire/async-tar), but it doesn't accept `AsyncWrite`)

If dependencies don't support `AsyncWrite`, we have to write a fair amount of glue code, which will be more complicated than using `Pipe`.

So, I'll pass on that this time.

Reply:
Alright, that's fair. Thanks for looking into it. I was sure that zip couldn't do it (and we already state in the help that generating a large zip will block the server) but I'm saddened to hear that there's nothing for tar. Alright, let's carry on then and revisit this in some while.